### ISSUE TITLE
Permute is inefficient on AMD CPUs

### ISSUE THREAD

**Issue by Const-me** [microsoft/DirectXMath#95]
TLDR: AVX `vpermilps` instruction is strictly worse than SSE `shufps` instruction.

This line
https://github.com/microsoft/DirectXMath/blob/master/Inc/DirectXMath.h#L170
slows down `XM_PERMUTE_PS` by a factor of 3-4 on AMD CPUs when AVX is enabled, with no benefits on Intel.

On recent Intel chips, both `shufps` and `vpermilps` have 1 cycle latency, 1 cycle throughput.

On AMD Ryzen, `shufps` has 1 cycle latency 0.5 cycles throughput, while vpermilps is 3-4 times slower, with latency 3 and throughput 2.

In addition, encoded shufps instruction is 1 byte shorter, 5 versus 6 bytes, i.e. there’s slight benefit even on Intel.

The only reason for `_mm_permute_ps` appears to be Intel’s [Xeon Phi](https://en.wikipedia.org/wiki/Xeon_Phi) where `vpermilps` can be faster with twice the throughput, but AFAIK DirectXMath doesn’t support that platform?

**Comment by walbourn** [comment]
Thanks for the detailed information.  As with many Intel instruction sets, they are often implemented with the assumption that future hardware will make them more optimal, but it seems that never materialized for this particular usage.

Ideally I'd be able to tell at compile-time if ``/favor:INTEL`` or ``/favor:AMD64`` was being used, but I don't see a value for it (only for ``/favor:ATOM``).

**Comment by walbourn** [comment]
Resolved in [this commit](https://github.com/microsoft/DirectXMath/commit/5008416aba655943b138a5e4541da8b8da8f5ecb)

**Comment by walbourn** [comment]
@Const-me : Which do you think is faster on AMD CPUs:

```
__m128  vTemp = _mm_broadcastss_ps(V);
```

or

```
__m128 vTemp = _mm_broadcastss_ps(V);
```

**Comment by Const-me** [comment]
Hi @walbourn.

On all recent Intel and AMD CPUs, `_mm_broadcastss_ps` has latency 3 cycles, throughput 1 cycle.
On all recent Intel CPUs, `_mm_shuffle_ps` has latency 1 cycle, throughout 1 cycle. On new AMD CPUs it’s even better, latency 1 cycle throughout 0.5 cycles i.e. it can run 2 of them per cycle. `_mm_shuffle_ps` is strictly faster.

You can find that data in the CHM I’ve generated, there: https://github.com/Const-me/IntelIntrinsics/ The lines with blue text in the performance table come from [intel.com](https://software.intel.com/sites/landingpage/IntrinsicsGuide/). The lines with black text are from [agner.org](https://www.agner.org/optimize/) (an independent academic researcher who benchmarked many CPUs including AMD ones).

BTW, there’re good use cases for that AVX2 instruction, `vbroadcastss`. One is the 256-bit version, `_mm256_broadcastss_ps` in C++. Another one is loading from memory as opposed to broadcasting from a register, `_mm_broadcast_ss` in C++, `shufps` can’t do that.

**Comment by walbourn** [comment]
Thanks, @Const-me, and that jives with my understanding. _mm_broadcastss_ps looks more like a 'might as well provide a uniform instruction set' option rather than super-useful, so that helps me determine if AVX2-specific paths for most cases are useful or not.

**Comment by walbourn** [comment]
Minor update in [this commit](https://github.com/microsoft/DirectXMath/commit/f68f691b369a88d0b43fcdb64b9436167d66b4e4)

**Comment by gm-matthew** [comment]
I know this is an old issue but for the record it is actually `vpermilps x,x,x/m`, the variable control version, that runs more slowly than `shufps x,x/m,i` on AMD Ryzen CPUs whereas the immediate control version, `vpermilps x,x/m,i` has the same performance, as also does `vbroadcastss x,x`.

According to https://www.agner.org/optimize/instruction_tables.pdf, on Zen 1-3 CPUs both `vpermilps x,x/m,i` and `vbroadcastss x,x` have a latency of 1 and reciprocal throughput of 0.5, the same as `shufps x,x/m,i`.

This means both `_mm_permute_ps()` and `_mm_broadcastss_ps()` have the same performance as `_mm_shuffle_ps()` on Ryzen CPUs.

It is `vpermilps x,x,x/m` and `vbroadcastss y,m32` that have the worse latency and reciprocal throughput figures that @Const-me claims; these correspond to `_mm_permutevar_ps()` and `_mm256_broadcast_ss()` respectively.
